File: app/dlf.py
#!/usr/bin/env python3

import requests
import json
import logging
from bs4 import BeautifulSoup, Tag
from sys import exit
from flask import abort

PREFIX = "https://www.deutschlandfunk.de/"

_WOCHENRUECKBLICK_URL = PREFIX + "nachrichten/wochenueberblick"

_LOGGER = logging.getLogger(__name__)

_NON_ARTICLES_URLS = [
    PREFIX + "nachrichten/nachlesen",
    PREFIX + "suche",
    PREFIX + "nachrichten/barrierefrei",
]


def download_wochenrueckblick() -> str:
    r = requests.get(_WOCHENRUECKBLICK_URL)
    if r.status_code != 200:
        print("Could not download the Wochenrückblick from Deutschlandfunk!")
        exit(1)
    return r.text


def _parse_partial_article(script: Tag) -> dict | None:
    j = json.loads(script["data-json"])["value"]

    if "__typename" not in j:
        if "data" in j and "newsByWeek" in j["data"]:
            # One extra element that is not an article
            return None
        _LOGGER.warning(
            "Found data element that does not fit scheme! See debug output for more information"
        )
        _LOGGER.debug(json.dumps(j))
        return None

    if j["__typename"] == "Teaser" and j["path"] not in _NON_ARTICLES_URLS:
        _LOGGER.info(f"  -> Found Article '{j['title']}'")
        if not j["path"].startswith(PREFIX):
            return None
        return {
            "key": j["path"][len(PREFIX) :],
            "title": j["title"],
            "teaserHeadline": j["teaserHeadline"],
            "teaserText": j["teasertext"],
            "date": j["firstPublicationDate"],
            "localeDate": j["dateLocalizedFormatted"],
        }
    else:
        _LOGGER.debug(f"Not a Teaser/News Article: {j}")
    return None


def parse_wochenrueckblick(wr: str) -> list[dict[str, str]]:
    soup = BeautifulSoup(wr, "lxml")
    main = soup.main

    if main is None:
        raise ValueError("<main> not found in the html test!")

    articles = []
    _LOGGER.info("Parsing Wochenrückblick...")
    for script in main.find_all("script"):
        article = _parse_partial_article(script)
        if article is not None:
            articles.append(article)
    return articles


def download_article(href: str) -> str:
    assert href is not None
    assert href.startswith("https://www.deutschlandfunk.de/")
    r = requests.get(href)
    if r.status_code != 200:
        raise ConnectionError(f"Could not download the article: {r.status_code}")
    return r.text


def parse_article(html: str) -> dict[str, str]:
    soup: Tag = BeautifulSoup(html, "lxml")

    # Parse Head for Metadata
    head: Tag = soup.head
    metadata: None | dict = None
    for script in head.find_all("script", {"class": "js-client-queries"}):
        metadata = _parse_partial_article(script)
        if metadata is not None:
            break

    # Parse content
    article: Tag = soup.find("article", {"class": "b-article"})
    header: Tag = article.header
    section: Tag = article.section
    assert section is not None

    figures: list[dict[str, str]] = []
    for figure in section.find_all("figure"):
        img: Tag = figure.img
        figures.append(
            {
                "image": {
                    "src": img["src"],
                    "alt": img["alt"],
                    "srcset": img["srcset"],
                    "title": img["title"],
                },
                "caption": (
                    figure.figcaption.text
                    if figure.figcaption is not None
                    else img["alt"]
                ),
            }
        )
        figure.decompose()
    for script in section.find_all("script"):
        script.decompose()

    return {
        "kicker": header.find("span", {"class": "headline-kicker"}).text,
        "title": header.find("span", {"class": "headline-title"}).text,
        "description": header.find("p", {"class": "article-header-description"}).text,
        "author": header.find("div", {"class": "article-header-author"}).span.text,
        "content": {
            "html": section.decode_contents().strip(),
            "plaintext": section.text.strip(),
        },
        "figures": figures,
        "key": metadata["key"],
        "date": metadata["date"],
        "localeDate": metadata["localeDate"],
    }

File: src/dlf.rs
use scraper::{Html, Selector};
use serde::{Deserialize, Serialize};

pub const PREFIX: &str = "https://www.deutschlandfunk.de/";
const WOCKENRUECKBLICK_URL: &str = "https://www.deutschlandfunk.de/nachrichten/wochenueberblick";
const NON_ARTICLE_URLS: [&str; 3] = ["nachrichten/nachlesen", "suche", "nachrichten/barrierefrei"];

macro_rules! select_one {
    ($parent:expr, $selector:expr) => {
        $parent
            .select(
                &::scraper::Selector::parse($selector)
                    .expect(&format!("Failed to parse the selector '{}'", $selector)),
            )
            .next()
            .expect(&format!(
                "No matching element found for selector '{}'",
                $selector
            ))
    };
}

#[derive(Deserialize, Serialize, Debug)]
pub struct PartialArticle {
    pub key: String,
    pub title: String,
    #[serde(rename = "teaserHeadline")]
    pub teaser_headline: Option<String>,
    #[serde(rename = "teasertext")]
    pub teaser_text: Option<String>,
    #[serde(rename = "firstPublicationDate")]
    pub date: String,
    #[serde(rename = "dateLocalizedFormatted")]
    pub locale_date: String,
}

fn parse_partial_article(json: &str) -> Result<Option<PartialArticle>, &'static str> {
    let json: serde_json::Value = serde_json::from_str(json).unwrap();
    let value = json.get("value").unwrap();
    match value.get("__typename") {
        None => {
            match json.get("data") {
                Some(data) => {
                    if data.get("newsByWeek").is_some() {
                        return Ok(None);
                    }
                    debug!(
                        "Found data element that does not fit scheme! See debug output for more information: {:?}",
                        &json
                    );
                }
                None =>
                    debug!(
                        "Found data element that does not fit scheme! See debug output for more information: {:?}",
                        &json
                    )
            }
            Ok(None)
        }
        Some(typename) => {
            let typename = typename.as_str().unwrap();
            if typename != "Teaser" {
                debug!("Not a teaser article");
                return Ok(None);
            }
            let path = value.get("path").unwrap().as_str().unwrap();

            for url in NON_ARTICLE_URLS.iter() {
                if path.starts_with(&format!("{}{}", PREFIX, url)) {
                    debug!("Not a news article!");
                    return Ok(None);
                }
            }

            // Add key from main json node to the value node
            let mut value = value.clone();
            value["key"] = path.strip_prefix(PREFIX).unwrap().into();
            let parsed: PartialArticle =
                serde_json::from_value(value.clone()).expect("Parsing the article went wrong!");
            Ok(Some(parsed))
        }
    }
}

pub async fn wochenrueckblick() -> Result<Vec<PartialArticle>, &'static str> {
    let body = reqwest::get(WOCKENRUECKBLICK_URL)
        .await
        .map_err(|_| "Could not retrieve the webpage")?
        .text()
        .await
        .map_err(|_| "Could not retrieve the webpage's text")?;

    let document = Html::parse_document(&body);
    let script_selector = Selector::parse("script").unwrap();

    let mut articles: Vec<PartialArticle> = vec![];
    let main = select_one!(document, "main");
    for script in main.select(&script_selector) {
        let parsed_article = parse_partial_article(script.attr("data-json").unwrap())?;
        if let Some(article) = parsed_article {
            articles.push(article);
        }
    }

    Ok(articles)
}

#[derive(Serialize, Debug)]
pub struct Article {
    pub kicker: String,
    pub title: String,
    pub description: String,
    pub html: String,
    pub plaintext: String,
    pub figures: Vec<Figure>,
    pub key: String,
    pub date: String,
    pub locale_date: String,
}

#[derive(Serialize, Debug)]
pub struct Figure {
    pub src: String,
    pub alt: String,
    pub srcset: String,
    pub title: String,
    pub caption: String,
}

pub async fn article(href: &str) -> Result<Article, &'static str> {
    if !href.starts_with(PREFIX) {
        return Err("Is not a DLF Article");
    }

    let response = reqwest::get(href).await.unwrap();
    if response.status() != 200 {
        return Err("Could not download article!");
    }

    let html = response.text().await.unwrap();
    let document = Html::parse_document(&html);

    // Parse head
    let head = select_one!(document, "head");
    let mut metadata: Option<PartialArticle> = None;
    for script in head.select(&Selector::parse("script.js-client-queries").unwrap()) {
        metadata = parse_partial_article(script.attr("data-json").unwrap()).unwrap();
        if metadata.is_some() {
            break;
        }
    }
    let metadata = metadata.unwrap();

    let article = select_one!(document, "article.b-article");
    let header = select_one!(article, "header");
    let section = select_one!(article, "section");

    let mut figures: Vec<Figure> = vec![];
    for figure in section.select(&Selector::parse("figure").unwrap()) {
        let image = select_one!(figure, "img");
        figures.push(Figure {
            src: String::from(image.attr("src").unwrap()),
            alt: String::from(image.attr("alt").unwrap()),
            srcset: String::from(image.attr("srcset").unwrap()),
            title: String::from(image.attr("title").unwrap()),
            caption: {
                if let Some(figcaption) = figure
                    .select(&Selector::parse("figcaption").unwrap())
                    .next()
                {
                    figcaption.text().collect()
                } else {
                    String::from(image.attr("alt").unwrap())
                }
            },
        });
    }

    Ok(Article {
        kicker: String::from(
            select_one!(header, "span.headline-kicker")
                .text()
                .collect::<String>()
                .trim(),
        ),
        title: String::from(
            select_one!(header, "span.headline-title")
                .text()
                .collect::<String>()
                .trim(),
        ),
        description: String::from(
            select_one!(header, "p.article-header-description")
                .text()
                .collect::<String>()
                .trim(),
        ),
        html: section.inner_html(),
        plaintext: section.text().collect(),
        figures,
        key: metadata.key,
        date: metadata.date,
        locale_date: metadata.locale_date,
    })
}

#[cfg(test)]
mod test {
    use super::*;

    #[tokio::test]
    async fn download_and_parse_wochenrueckblick() {
        wochenrueckblick().await.unwrap();
    }

    #[tokio::test]
    async fn download_and_parse_article() {
        article("https://www.deutschlandfunk.de/bundesregierung-will-asylrecht-weiter-verschaerfen-100.html").await.unwrap();
    }
}
